### DISCUSSION TITLE
Help me figure out the blocks!!!

### DISCUSSION THREAD

**Discussion by Quasarrs** [mozman/ezdxf#945] [Q&A]
Dear developer!!!, please tell me in version 1.1.1 of the library there are only 2 methods for inserting a block into a model???
1. `add_blockref(name: str, insert: UVec, dxfattribs=None)→ Insert`
2. `add_auto_blockref(name: str, insert: UVec, values: dict[str, str], dxfattribs=None)→ Insert`

in the first method I simply insert a block, and in the second I can immediately set the attribute values ​​to be passed if they are in the block.
Why, if the created block with multi-line text, when using the second method, turns into an attribute with single-line text???
Now how can this be solved?? if I strictly want to create blocks in a CAD program with multi-line text, and using your library they need to be placed and filled with attributes.
Or is this not possible now?

**Comment by mozman** [comment]
Hi!

Again #942: Multi-line attributes were introduced in DXF R2018 and are not documented. Avoid them if you can.

Multi-line attributes are not supported by the methods `Layout.add_auto_blockref()` and `Insert.add_auto_attribs()`, these methods create only single line attributes.

There is one [example](https://github.com/mozman/ezdxf/blob/master/examples/blocks/insert_with_mtext_attrib.py) how to add multi-line attributes to INSERT entities: 

```Python
import ezdxf

def main():
    doc = ezdxf.new(dxfversion="R2018")
    # create a new block definition:
    block = doc.blocks.new("TEST")
    block.add_circle((0, 0), 1)

    msp = doc.modelspace()
    # create a block reference in modelspace:
    insert = msp.add_blockref("TEST", (0, 0))
    # add a single line attribute to the block reference:
    attrib = insert.add_attrib("ATT0", "dummy")
    # create a MTEXT entity:
    m_text = msp.add_mtext("test1\ntest2\ntest3")
    m_text.set_location((5, 5))
    # embed the MTEXT entity into the attribute:
    attrib.embed_mtext(m_text)

    doc.saveas("insert_with_mtext_attribute.dxf")


if __name__ == "__main__":
    main()
```

You have to do everything manually:
- create a new ATTRIB entity
- create a new MTEXT entity
- embed the MTEXT entity into the ATTRIB entity

An ATTRIB entity does not require an ATTDEF entity in the BLOCK definition, the ATTDEF is just a template to create new ATTRIB entities.

**Reply by Quasarrs** [reply]
This is not exactly what I want to get(....

The point is that in your example you create an attribute inside the block in the code and show where it will be located.
`m_text.set_location((5, 5))`

I really need something a little different. The block with the attributes was drawn in a CAD program, I just want to use Python to place them where I need them and set the values ​​of the multi-line attributes.

Now I can’t solve this problem using your library?


Maybe I'll try to describe it in more detail:

Imagine that you have been given a .dwg file in which there are already blocks inside the drawing with multi-line text attributes. The task is to place these blocks where necessary and at the same time transfer different text values ​​to these attributes

**Comment by mozman** [comment]
Editing the embedded `MTEXT` entity is not supported directly.

Creating the source drawing:
![doc0](https://github.com/mozman/ezdxf/assets/4230966/4152ef25-be13-4a28-a69a-bbb034270096)

```Python
import pathlib
import ezdxf
from ezdxf.entities import Insert
from ezdxf.math import Matrix44

CWD = pathlib.Path("~/Desktop/Outbox").expanduser()
if not CWD.exists():
    CWD = pathlib.Path(".")

DOC_0 = CWD / "source_mtext_attribute.dxf"
DOC_1 = CWD / "edit_mtext_attribute.dxf"
DOC_2 = CWD / "move_insert_and_attrib.dxf"
DOC_3 = CWD / "move_attribute_without_insert.dxf"


def make_doc():
    doc = ezdxf.new(dxfversion="R2018")
    block = doc.blocks.new("TEST")
    block.add_circle((0, 0), 1)
    msp = doc.modelspace()
    insert = msp.add_blockref("TEST", (0, 0))
    attrib = insert.add_attrib("ATT0", "dummy")
    m_text = msp.add_mtext("test1\ntest2\ntest3")
    m_text.set_location((5, 5))
    attrib.embed_mtext(m_text)
    doc.saveas(DOC_0)
```

Editing the embedded `MTEXT` entity:
![doc1](https://github.com/mozman/ezdxf/assets/4230966/84858c5c-db45-4073-bf81-390e382ab7d7)

```Python
def edit_multiline_attrib():
    doc = ezdxf.readfile(DOC_0)
    insert: Insert = doc.query("INSERT").first
    attrib = insert.get_attrib("ATT0")
    if attrib.has_embedded_mtext_entity:
        # extract the embedded MTEXT entity
        mtext = attrib.virtual_mtext_entity()
        # modify the content
        mtext.text += "\nline4"  # text is not in the DXF namespace!
        # set absolut location
        mtext.dxf.insert = (5, 5)
        # copy modified MTEXT content to ATTRIB
        attrib.set_mtext(mtext)
        doc.saveas(DOC_1)
```

Transform the INSERT and the ATTRIB entity at once - this works with the next release v1.1.2, because the current release has a bug:
![doc2](https://github.com/mozman/ezdxf/assets/4230966/746c3408-780d-4a38-876b-578135f875f8)

```Python
def move_insert_and_attrib():
    doc = ezdxf.readfile(DOC_0)
    insert: Insert = doc.query("INSERT").first
    insert.transform(Matrix44.translate(0, -5, 0))
    doc.saveas(DOC_2)
```

Transform only the ATTRIB entity - this works with the next release v1.1.2, because the current release has a bug:
![doc3](https://github.com/mozman/ezdxf/assets/4230966/3d510e10-3404-48d6-bc1d-3cbdfc9243e1)
```Python
def move_attrib_only():
    doc = ezdxf.readfile(DOC_0)
    insert: Insert = doc.query("INSERT").first
    attrib = insert.get_attrib("ATT0")
    attrib.transform(Matrix44.translate(0, 5, 0))
    doc.saveas(DOC_3)
```

Executing the code:
```Python
if __name__ == "__main__":
    make_doc()
    edit_multiline_attrib()
    move_insert_and_attrib()
    move_attrib_only()
```

Source code: 
[script.zip](https://github.com/mozman/ezdxf/files/13198199/script.zip)

**Comment by Quasarrs** [comment]
Mozman, thank you so much for showing in such detail how you can work with these blocks using other means.
I don’t quite understand what will change in 1.1.2 “Transform the INSERT and the ATTRIB entity at once”

**Comment by Quasarrs** [comment]
I looked at your example again.. The difficulty I am experiencing...
I need to uncopy one block.
 When searching for an attribute with Mtext, the search is performed throughout the entire doc. It turns out that when copying, you need to look for attributes in the block?
Here's an example added

1. we will assume that this is already created DOC_0
![image](https://github.com/mozman/ezdxf/assets/145274914/4dc10152-4b66-4db0-837f-61f7be59aa60)
2.
As I understand it, reading DOC_0, I won’t be able to create copies like in the photo below? Only using the add_blockref method which breaks multiline text 
![image](https://github.com/mozman/ezdxf/assets/145274914/90b13b9b-a7e2-4831-9140-62954f6e90b6)
